File: MergeSortedArray.java
public class MergeSortedArray {
    /*
        - Ý tưởng giải bài toán
        1. Bắt đầu từ cuối của mảng nums1 (phần tử hợp nhất cuối cùng).
        2. So sánh các phần tử từ cuối của nums1 và nums2.
        3. Chèn phần tử lớn hơn vào cuối nums1.
        4. Tiếp tục giảm chỉ số và lặp lại quá trình đến khi duyệt hết các phần tử.
     */
    public static void merge(int[] a, int m, int[] b, int n) {
        int i = m - 1;
        int j = n - 1;
        int k = m + n - 1;

        while (i >= 0 || j >= 0) {
            if (i >= 0 && j >= 0) {
                if (a[i] >= b[j]) {
                    a[k--] = a[i--];
                } else {
                    a[k--] = b[j--];
                }
            } else if (i >= 0) {
                a[k--] = a[i--];
            } else {
                a[k--] = b[j--];
            }
        }
    }

    public static void main(String[] args) {
        int[] a = {1, 2, 3, 4, 0, 0, 0, 0};
        int m = 4;
        int[] b = {2, 5, 6, 9};
        int n = 4;
        merge(a, m, b, n);
        for (int i = 0; i < a.length; i++) {
            System.out.print(a[i] + " ");
        }
    }
}

File: Array/MergeSortedArray.js
/**
 * @param {number[]} nums1
 * @param {number} m
 * @param {number[]} nums2
 * @param {number} n
 * @return {void} Do not return anything, modify nums1 in-place instead.
 */
var merge = function (nums1, m, nums2, n) {
  // Khởi tạo các con trỏ
  let p1 = m - 1; // Chỉ số cuối của phần tử hợp lệ trong nums1
  let p2 = n - 1; // Chỉ số cuối của nums2
  let p = m + n - 1; // Chỉ số cuối của nums1 (vị trí để điền)

  // Khi cả hai mảng còn phần tử
  while (p1 >= 0 && p2 >= 0) {
    if (nums1[p1] > nums2[p2]) {
      nums1[p] = nums1[p1]; // Điền phần tử lớn hơn từ nums1
      p1--; // Giảm chỉ số nums1
    } else {
      nums1[p] = nums2[p2]; // Điền phần tử lớn hơn từ nums2
      p2--; // Giảm chỉ số nums2
    }
    p--; // Giảm chỉ số vị trí điền
  }

  // Nếu nums2 còn phần tử, điền hết vào nums1
  while (p2 >= 0) {
    nums1[p] = nums2[p2];
    p2--;
    p--;
  }
};

// Ví dụ sử dụng
let nums1 = [1, 2, 3, 0, 0, 0];
let m = 3;
let nums2 = [2, 5, 6];
let n = 3;
merge(nums1, m, nums2, n);
console.log(nums1); // Kết quả: [1, 2, 2, 3, 5, 6]

